kernel/src/table_changes/log_replay.rs

```diff
@@ -0,0 +1,847 @@
+        schema: SchemaRef,
+    ) -> Self {
+        Self {
+            timestamp: commit_file.location.last_modified,
+            commit_file,
+            json_handler,
+            filter,
+            has_cdc_action: Default::default(),
+            remove_dvs: Default::default(),
+            expression_handler,
+            schema,
+        }
+    }
+    fn prepare_phase(&mut self) -> DeltaResult<()> {
+        let schema = PreparePhaseVisitor::schema()?;
+        let action_iter = self.json_handler.read_json_files(
+            &[self.commit_file.location.clone()],
+            schema,
+            None,
+        )?;
+
+        let mut add_paths = HashSet::new();
+        for actions in action_iter {
+            let actions = actions?;
+
+            // Apply data skipping to get back a selection vector for actions that passed skipping.
+            // We start our selection vector based on what was filtered. We will add to this vector
+            // below if a file has been removed. Note: None implies all files passed data skipping.
+            let selection_vector = match &self.filter {
+                Some(filter) => filter.apply(actions.as_ref())?,
+                None => vec![true; actions.len()],
+            };
+
+            let mut visitor = PreparePhaseVisitor::new(self, selection_vector, &mut add_paths);
+            visitor.visit_rows_of(actions.as_ref())?;
+
+            if let Some(protocol) = visitor.protocol {
+                protocol.ensure_read_supported()?;
+            }
+            if let Some((schema, configuration)) = visitor.metadata_info {
+                let schema: StructType = serde_json::from_str(&schema)?;
+                require!(
+                    self.schema.as_ref() == &schema,
+                    Error::change_data_feed_incompatible_schema(&self.schema, &schema)
+                );
+
+                let table_properties = TableProperties::from(configuration);
+                require!(
+                    table_properties.enable_change_data_feed.unwrap_or(false),
+                    Error::change_data_feed_unsupported(self.commit_file.version)
+                )
+            }
+        }
+
+        // We resolve the remove deletion vector map after visiting the entire commit.
+        if self.has_cdc_action {
+            self.remove_dvs.clear();
+        } else {
+            // The only (path, deletion_vector) pairs we must track are ones whose path is the
+            // same as an `add` action.
+            self.remove_dvs
+                .retain(|rm_path, _| add_paths.contains(rm_path));
+        }
+        Ok(())
+    }
+
+    fn into_scan_batches(
+        self,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<TableChangesScanData>>> {
+        let Self {
+            has_cdc_action,
+            remove_dvs,
+            commit_file,
+            json_handler,
+            timestamp: _,
+            filter,
+            expression_handler,
+            schema: _,
```

Comment:
Correct. I was considering doing some kind of type state for the two phases, but I wanted to avoid complicated designs to speed things up. Do you think the phases should be split up?

Reply:
I think _not_ splitting up is making the PR _more_ complicated, and that things would be cleaner/easier to follow if we split it.

Reply:
Yeah you're right. You'd requested to move the prepare phase into a standalone function. This ended up trimming away all the unused fields from LogReplay.